go-controller/pkg/node/node_ip_handler_linux.go

```diff
@@ -0,0 +1,149 @@
+)
+
+type addressManager struct {
+	addresses      sets.String
+	nodeAnnotator  kube.Annotator
+	mgmtPortConfig *managementPortConfig
+	mu             sync.Mutex
+}
+
+// initializes a new address manager which will hold all the IPs on a node
+func newAddressManager(nodeAnnotator kube.Annotator, config *managementPortConfig) *addressManager {
+	mgr := &addressManager{
+		addresses:      sets.NewString(),
+		nodeAnnotator:  nodeAnnotator,
+		mgmtPortConfig: config,
+		mu:             sync.Mutex{},
+	}
+
+	klog.Info("Address manager created with IPs: %s", mgr.addresses.List())
+	return mgr
+}
+
+// updates the address manager with a new IP
+// returns true if there was an update
+func (c *addressManager) addAddr(ip net.IP) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.addresses.Has(ip.String()) && c.isValidNodeIP(ip) {
+		klog.V(5).Infof("Adding IP: %s, to node IP manager", ip)
+		c.addresses.Insert(ip.String())
+		return true
+	}
+
+	return false
+}
+
+// removes IP from address manager
+// returns true if there was an update
+func (c *addressManager) delAddr(ip net.IP) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.addresses.Has(ip.String()) && c.isValidNodeIP(ip) {
+		klog.V(5).Infof("Removing IP: %s, from node IP manager", ip)
+		c.addresses.Delete(ip.String())
+		return true
+	}
+
+	return false
+}
+
+func (c *addressManager) Run(stopChan <-chan struct{}) {
+	addrChan := make(chan netlink.AddrUpdate)
+	if err := netlink.AddrSubscribe(addrChan, stopChan); err != nil {
+		klog.Fatalf("Unable to run Address Manager, error during netlink subscribe: %v", err)
+	}
+	go func() {
+		for {
+			select {
+			case a := <-addrChan:
+				if a.NewAddr {
+					if c.addAddr(a.LinkAddress.IP) {
+						if err := util.SetNodeHostAddresses(c.nodeAnnotator, c.addresses); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+							continue
+						}
+						if err := c.nodeAnnotator.Run(); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+						}
+					}
+				} else {
+					if c.delAddr(a.LinkAddress.IP) {
+						if err := util.SetNodeHostAddresses(c.nodeAnnotator, c.addresses); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+							continue
+						}
+						if err := c.nodeAnnotator.Run(); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+						}
+					}
+				}
+			case <-stopChan:
+				close(addrChan)
+				return
+			}
+		}
+	}()
+
+	addrs, err := netlink.AddrList(nil, netlink.FAMILY_ALL)
+	if err != nil {
+		klog.Fatalf("Failed to run Node IP Manager: unable list all IPs on the node, error: %v", err)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, addr := range addrs {
+		if c.isValidNodeIP(addr.IP) {
+			c.addresses.Insert(addr.IP.String())
+		}
+	}
+	if err := util.SetNodeHostAddresses(c.nodeAnnotator, c.addresses); err != nil {
+		klog.Errorf("Failed to set node annotations: %v", err)
+	} else {
+		if err := c.nodeAnnotator.Run(); err != nil {
```

Comment:
@trozet also the node annotator is not thread-safe; it doesn't lock. So this may not work so well with the code above in the goroutine.

Reply:
i'm going to have to think about this some more. Even nodeAnnotator.Set is not thread safe as it tries to write to a map. Maybe I should just bite the bullet and make the annotator thread safe.

Reply:
It's not too complicated, either move the goroutine after the initial annotation set, or use a channel to make the goroutine block before it runs the for() loop that you close right after doing the initial set.

Reply:
Well other things may run and use the node annotator in the main thread so I think it might be better to just make the annotator safe. Helps future proof us

Reply:
ah, I'd thought you'd created a new one. So yeah, should just bit the bullet and maek it safe.